core/slice.go

```diff
@@ -354,33 +348,178 @@ func (sl *Slice) Append(header *types.WorkObject, domTerminus common.Hash, domOr
+					if len(etx.Data()) > 1 {
+						slipAmount = new(big.Int).SetBytes(etx.Data()[:2])
+						if slipAmount.Cmp(params.MaxSlip) > 0 {
+							slipAmount = new(big.Int).Set(params.MaxSlip)
+						}
+						if slipAmount.Cmp(params.MinSlip) < 0 {
+							slipAmount = new(big.Int).Set(params.MinSlip)
+						}
+					}
+
+					// Apply the cubic discount based on the current tempActualConversionAmountInQuai
+					discountedConversionAmount := misc.ApplyCubicDiscount(block.ConversionFlowAmount(), tempActualConversionAmountInQuai)
+					discountedConversionAmountInInt, _ := discountedConversionAmount.Int(nil)
+
+					// Conversions going against the direction of the exchange
+					// rate adjustment dont get any k quai discount applied
+					conversionAmountAfterKQuaiDiscount := new(big.Int).Mul(discountedConversionAmountInInt, new(big.Int).Sub(big.NewInt(params.KQuaiDiscountMultiplier), kQuaiDiscount))
+					conversionAmountAfterKQuaiDiscount = new(big.Int).Div(conversionAmountAfterKQuaiDiscount, big.NewInt(params.KQuaiDiscountMultiplier))
 
 					// Apply the conversion flow discount to all transactions
 					value = new(big.Int).Mul(value, discountedConversionAmountInInt)
-					value = new(big.Int).Div(value, conversionAmountInQuai)
+					value = new(big.Int).Div(value, tempActualConversionAmountInQuai)
 
+					// ten percent original value is calculated so that in case
+					// the slip is more than this it can be reset
+					tenPercentOriginalValue := new(big.Int).Mul(originalValue, common.Big10)
+					tenPercentOriginalValue = new(big.Int).Div(tenPercentOriginalValue, common.Big100)
+
+					// amount after max slip is calculated based on the
+					// specified slip so that if the final value exceeds this,
+					// the transaction can be reverted
+					amountAfterMaxSlip := new(big.Int).Mul(originalValue, new(big.Int).Sub(params.SlipAmountRange, slipAmount))
+					amountAfterMaxSlip = new(big.Int).Div(amountAfterMaxSlip, params.SlipAmountRange)
+
+					// If to is in Qi, convert the value into Qi
+					if etx.To().IsInQiLedgerScope() {
+						// Apply the k quai discount only if the exchange rate is increasing
+						if exchangeRateIncreasing && discountedConversionAmountInInt.Cmp(common.Big0) != 0 {
+							value = new(big.Int).Mul(value, conversionAmountAfterKQuaiDiscount)
+							value = new(big.Int).Div(value, discountedConversionAmountInInt)
+						}
+					}
+
+					// If To is in Quai, convert the value into Quai
+					if etx.To().IsInQuaiLedgerScope() {
+						// Apply the k quai discount only if the exchange rate is decreasing
+						if !exchangeRateIncreasing && discountedConversionAmountInInt.Cmp(common.Big0) != 0 {
+							value = new(big.Int).Mul(value, conversionAmountAfterKQuaiDiscount)
+							value = new(big.Int).Div(value, discountedConversionAmountInInt)
+						}
+					}
+
+					// If the value is less than the ten percent of the
+					// original, reset it to 10%
+					if value.Cmp(tenPercentOriginalValue) < 0 {
```

Comment:
Since you have this, you don't need to set the max slip to 90% right?

Reply:
that is true, the max slip was more for not getting negative value, though i agree here, lets not touch it as its tested, we can remove it in the future